azurerm/resource_arm_analysis_services_server.go

```diff
@@ -0,0 +1,406 @@
+	adminUsers := expandAnalysisServicesServerAdminUsers(d)
+
+	serverProperties := analysisservices.ServerProperties{AsAdministrators: adminUsers}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	if querypoolConnectionMode, ok := d.GetOk("querypool_connection_mode"); ok {
+		serverProperties.QuerypoolConnectionMode = analysisservices.ConnectionMode(querypoolConnectionMode.(string))
+	}
+
+	return &serverProperties
+}
+
+func expandAnalysisServicesServerMutableProperties(d *schema.ResourceData) *analysisservices.ServerMutableProperties {
+	adminUsers := expandAnalysisServicesServerAdminUsers(d)
+
+	serverProperties := analysisservices.ServerMutableProperties{AsAdministrators: adminUsers}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	serverProperties.QuerypoolConnectionMode = analysisservices.ConnectionMode(d.Get("querypool_connection_mode").(string))
+
+	return &serverProperties
+}
+
+func expandAnalysisServicesServerAdminUsers(d *schema.ResourceData) *analysisservices.ServerAdministrators {
+	adminUsers := d.Get("admin_users").(*schema.Set)
+	members := make([]string, 0)
+
+	for _, admin := range adminUsers.List() {
+		if adm, ok := admin.(string); ok {
+			members = append(members, adm)
+		}
+	}
+
+	return &analysisservices.ServerAdministrators{Members: &members}
+}
+
+func expandAnalysisServicesServerFirewallSettings(d *schema.ResourceData) *analysisservices.IPv4FirewallSettings {
+	firewallSettings := analysisservices.IPv4FirewallSettings{}
+
+	if enablePowerBi, exists := d.GetOkExists("enable_power_bi_service"); exists {
+		firewallSettings.EnablePowerBIService = utils.Bool(enablePowerBi.(bool))
+	} else {
+		firewallSettings.EnablePowerBIService = utils.Bool(false)
+	}
+
+	firewallRules := d.Get("ipv4_firewall_rule").([]interface{})
+
+	fwRules := make([]analysisservices.IPv4FirewallRule, len(firewallRules))
+	for i, v := range firewallRules {
+		fwRule := v.(map[string]interface{})
+		fwRules[i] = analysisservices.IPv4FirewallRule{
+			FirewallRuleName: utils.String(fwRule["name"].(string)),
+			RangeStart:       utils.String(fwRule["range_start"].(string)),
+			RangeEnd:         utils.String(fwRule["range_end"].(string)),
+		}
+	}
+	firewallSettings.FirewallRules = &fwRules
+
+	return &firewallSettings
+}
+
+func flattenAnalysisServicesServerFirewallSettings(serverProperties *analysisservices.ServerProperties) (enablePowerBi *bool, fwRules []interface{}) {
+	if serverProperties.IPV4FirewallSettings == nil {
+		return utils.Bool(false), make([]interface{}, 0)
+	}
+
+	firewallSettings := serverProperties.IPV4FirewallSettings
+
+	enablePowerBi = utils.Bool(false)
+	if firewallSettings.EnablePowerBIService != nil {
+		enablePowerBi = firewallSettings.EnablePowerBIService
+	}
+
+	fwRules = make([]interface{}, 0)
+	for _, fwRule := range *firewallSettings.FirewallRules {
```

Comment:
if `firewallSettings.FirewallRules` is nil (for example, a bad server response) there's a potential crash here; as such could we make this:

```suggestion
	if firewallSettings.FirewallRules != nil {
		for _, fwRule := range *firewallSettings.FirewallRules {
		...
	}
```